data/fin.lean

```diff
@@ -24,6 +21,85 @@ by cases j; simp [fin.succ]
 @[simp] lemma pred_val (j : fin (n+1)) (h : j ≠ 0) : (j.pred h).val = j.val.pred :=
 by cases j; simp [fin.pred]
 
+@[simp] lemma fin.succ_pred (i : fin (n+1)) (H : i ≠ 0) :
+  (i.pred H).succ = i :=
+begin
+  apply fin.eq_of_veq,
+  cases i with i hi,
+  cases i,
+  { exfalso, apply H, apply fin.eq_of_veq, refl },
+  refl
+end
+
+@[simp] lemma fin.pred_succ (i : fin n) (H : i.succ ≠ 0) :
+  i.succ.pred H = i :=
+by cases i; refl
+
+/-- Embedding of `fin n` in `fin (n+1)` -/
+def raise (k : fin n) : fin (n + 1) := ⟨val k, lt_succ_of_lt (is_lt k)⟩
+
+@[simp] lemma fin.raise_val (k : fin n) : k.raise.val = k.val := rfl
+
+def lower : Π i : fin (n+1), i.1 < n → fin n := λ i h, ⟨i.1, h⟩
+
+@[simp] lemma lower_val (k : fin (n+1)) (H : k.1 < n) : (k.lower H).val = k.val := rfl
+
+def ascend (pivot : fin (n+1)) : Π i : fin n, fin (n+1) :=
+λ i, if i.1 < pivot.1 then i.raise else i.succ
```

Comment:
Why `Π i : ` and not use the `i`? You can just move this before the `:`. 
You can do this everywhere, at `lower`, `descend`, ...